src/Grpc.Net.Client/Internal/Configuration/ConfigProperty.cs

```diff
@@ -37,13 +37,23 @@ public ConfigProperty(Func<TInner?, TValue?> valueFactory, string key)
         {
             if (_value == null)
             {
-                var innerValue = inner.GetValue<TInner>(_key);
-                _value = _valueFactory(innerValue);
-
-                if (_value != null && innerValue == null)
+                // Multiple threads can get a property at the same time. We want this to be safe.
+                // Because a value could be lazyily initialized, lock to ensure multiple threads
+                // don't try to update the underlying dictionary at the same time.
+                lock (this)
                 {
-                    // Set newly created value
-                    SetValue(inner, _value);
+                    // Double-check locking.
+                    if (_value == null)
+                    {
+                        var innerValue = inner.GetValue<TInner>(_key);
+                        _value = _valueFactory(innerValue);
+
+                        if (_value != null && innerValue == null)
+                        {
+                            // Set newly created value
+                            SetValue(inner, _value);
```

Comment:
Should the SetValue function also lock since it's public and called by other code?

Reply:
`ConfigProperty` and `IConfigValue` are both internal so I think it's okay. If this was ever exposed anywhere, locking on `this` would be bad. But unless it's being exposed as `object`, I don't see how anyone would easily get a reference.

Reply:
> Should the SetValue function also lock since it's public and called by other code?

The goal is to make the type thread-safe for reads, e.g. multiple threads can do this without error:

```csharp
if (serviceConfig.LoadBalancingConfigs.Count > 0)
{
    // ....
}
```

If someone calls SetValue value by setting a property on another thread, e.g. `serviceConfig.LoadBalancingConfigs = new List<LoadBalancingConfig>()` then an error is fine.